packages/firebase_dynamic_links/ios/Classes/FirebaseDynamicLinksPlugin.m

```diff
@@ -0,0 +1,219 @@
+#import "FirebaseDynamicLinksPlugin.h"
+
+#import "Firebase/Firebase.h"
+
+@interface NSError (FlutterError)
+@property(readonly, nonatomic) FlutterError *flutterError;
+@end
+
+@implementation NSError (FlutterError)
+- (FlutterError *)flutterError {
+  return [FlutterError errorWithCode:[NSString stringWithFormat:@"Error %d", (int)self.code]
+                             message:self.domain
+                             details:self.localizedDescription];
+}
+@end
+
+@implementation FLTFirebaseDynamicLinksPlugin
++ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
+  FlutterMethodChannel *channel =
+      [FlutterMethodChannel methodChannelWithName:@"plugins.flutter.io/firebase_dynamic_links"
+                                  binaryMessenger:[registrar messenger]];
+  FLTFirebaseDynamicLinksPlugin *instance = [[FLTFirebaseDynamicLinksPlugin alloc] init];
+  [registrar addMethodCallDelegate:instance channel:channel];
+}
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    if (![FIRApp defaultApp]) {
+      [FIRApp configure];
+    }
+  }
+  return self;
+}
+
+- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
+  if ([@"DynamicLinkComponents#url" isEqualToString:call.method]) {
+    FIRDynamicLinkComponents *components = [self setupParameters:call.arguments];
+    result([components.url absoluteString]);
+  } else if ([@"DynamicLinkComponents#shortLink" isEqualToString:call.method]) {
+    FIRDynamicLinkComponents *components = [self setupParameters:call.arguments];
+    [components shortenWithCompletion:[self createShortLinkCompletion:result]];
+  } else if ([@"DynamicLinkComponents#shortenUrl" isEqualToString:call.method]) {
+    FIRDynamicLinkComponentsOptions *options = [self setupOptions:call.arguments];
+    NSURL *url = [NSURL URLWithString:call.arguments[@"url"]];
+    [FIRDynamicLinkComponents shortenURL:url
+                                 options:options
+                              completion:[self createShortLinkCompletion:result]];
+  } else {
+    result(FlutterMethodNotImplemented);
+  }
+}
+
+- (FIRDynamicLinkShortenerCompletion)createShortLinkCompletion:(FlutterResult)result {
+  return ^(NSURL *_Nullable shortURL, NSArray *_Nullable warnings, NSError *_Nullable error) {
+    if (error) {
+      result([error flutterError]);
+    } else {
+      result(@{@"url" : [shortURL absoluteString], @"warnings" : warnings});
+    }
+  };
+}
+
+- (FIRDynamicLinkComponentsOptions *)setupOptions:(NSDictionary *)arguments {
+  FIRDynamicLinkComponentsOptions *options;
+  if (![arguments[@"dynamicLinkComponentsOptions"] isEqual:[NSNull null]]) {
```

Comment:
I get the warning
```
Comparison of distinct pointer types ('NSString *' and 'NSNull * _Nonnull')
```
when converting
```
if (![affiliateToken != [NSNull null]])
```
to
```
if (affiliateToken != [NSNull null])
```

Is this what you were wanting me to change to?

Reply:
Ah, right. This is because the static type of `affiliateToken` is not `id`. I've seen people use something like
```
NSString* foo = ...
if (foo != (id)[NSNull null]) { ... }
```
[elsewhere](http://blog.alwawee.com/2014/11/19/check-nsdictionary-key-check-value-key-nil/). That may not strike you as much prettier than `![foo isEqual: [NSNull]]`.
Your call.